### PR TITLE
Reimagine Sharing: Clip playback

### PR DESCRIPTION
**Pull request by bjtitus**
| 📘 Part of: #1910 |
|:---:|

Adds clip playback via a new `ClipPlaybackManager` which uses a separate `AVPlayer` instance.

https://github.com/user-attachments/assets/ceabce5f-2162-484e-8944-b214dff11c63

### Known Issues

* The playback head disappears behind the trailing trim handle
* The "Audio waveform" doesn't look right at the zoom level
* The playback position indicator is not as smooth as it might be
* Trim handles are slightly sluggish during playback, especially on older devices

## To test

> [!NOTE]
> Enable the `newSharing` feature flag to test this

### Playback

* Play an episode
* Navigate to the player and then the "Share" button in the action bar
* Tap "Clip"
* Verify that tapping the Play button plays back the audio for the clip
* Tapping Pause should stop audio playback
* Moving the playhead should change the current play time when ending the gesture
* Moving the Trim Handles should change the start/stop point of playback
* Upon reaching the end, the clip should seek back to the beginning and pause

### Sharing 

* Tap the "Clip" button
* Tap "Copy Link"
* Verify that the link which was created mirrors the clip you expected to create. This clip should play in the web page.
* Try testing with changing the range of the clip in the editor

### Dismiss Clip Editor

* Navigate back to the Player
* Ensure that the episode playback position remains where it was when you began sharing the clip

### Episode Detail

* Navigate to an Episode Detail page for an episode which isn't playing but which you have played a bit of
* Tap the share button in the upper right
* Tap "Clip"
* Verify that the clip editor page comes up
* Verify that playback works but doesn't affect the currently playing episode

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/56">7.70</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: podcasts/Sharing/Clip/ClipPlaybackManager.swift [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    @ObservedObject var clipTime: ClipTime = ClipTime(start: 0, end: 0)
+
+    func play(episode: BaseEpisode, clipTime: ObservedObject<ClipTime>) {
+        if avPlayer != nil {
+            stop()
+            avPlayer = nil
+        }
+
+        PlaybackManager.shared.pause()
```
Comment (leandroalonso) [inline podcasts/Sharing/Clip/ClipPlaybackManager.swift]: Shall we inject this?

**Review by leandroalonso — CHANGES_REQUESTED** [review]
@bjtitus tested and it works as described, there are a few things worth mentioning:

1. When tapping play and closing the screen, playback don't stop. We need to ensure that leaving the screen stops the playback;
2. The playhead can be out of bounds with the trimmer: is that expected? Taking as an example the Photos app: the playhead is always contained inside the trimmer bounds.
3. I wonder if we should follow the Photos behavior: when moving the trimmer, any playback stops and the playhead disappears. After the trimming is done, the playhead re-appears at the beginning of the trimmed part. This should also fix the "Trim handles are slightly sluggish during playback, especially on older devices" known issue
4. When moving the playhead for a streaming episode it moves back to the previous position while the seek is being performed (it would be better to stay at the moved position)

None of those are blockers except (1)

**Comment by bjtitus** [comment]
**1**

> When tapping play and closing the screen, playback don't stop

✅ Fixed

**2**
> The playhead can be out of bounds with the trimmer

Yep, I have code for this but just haven't pushed it yet. I'll do that as a separate PR.

**3**
> I wonder if we should follow the Photos behavior: when moving the trimmer, any playback stops and the playhead disappears.

This is an interesting idea. I'll take a look at this alongside 2. It certainly would help a bit with performance, which I think is mostly a bunch of SwiftUI updates triggered by the moving playhead. Would this make it harder to line up the exact clip position, though?

**4**
> When moving the playhead for a streaming episode it moves back to the previous position while the seek is being performed

Are you talking about the quick jump? I was looking a bit and I think I will have to delay the update while seek is happening.

https://github.com/user-attachments/assets/1e5fa956-64d9-4eaf-b90e-6a8b82055a84

**Comment by bjtitus** [comment]
@leandroalonso I added the seek ignoring logic for number 4 in https://github.com/Automattic/pocket-casts-ios/pull/1947/commits/601a27fa230d88ef3124b69a4f372b78dbcc7500

**Review by leandroalonso — APPROVED** [review]
✅ Tested closing the screen while playback is ongoing
✅ Tested moving the playhead while streaming

**Pull request merged**